src/libs/TagsOptionsListUtils.ts

```diff
@@ -0,0 +1,170 @@
+        const selectedSearchTags = selectedTagsWithDisabledState.filter((tag) => PolicyUtils.getCleanedTagName(tag.name.toLowerCase()).includes(searchValue.toLowerCase()));
+        const tagsForSearch = [...selectedSearchTags, ...enabledSearchTags];
+
+        tagSections.push({
+            // "Search" section
+            title: '',
+            shouldShow: true,
+            data: getTagsOptions(tagsForSearch, selectedOptions),
+        });
+
+        return tagSections;
+    }
+
+    if (numberOfTags < CONST.TAG_LIST_THRESHOLD) {
+        tagSections.push({
+            // "All" section when items amount less than the threshold
+            title: '',
+            shouldShow: false,
+            data: getTagsOptions([...selectedTagsWithDisabledState, ...enabledTagsWithoutSelectedOptions], selectedOptions),
+        });
+
+        return tagSections;
+    }
+
+    const filteredRecentlyUsedTags = recentlyUsedTags
+        .filter((recentlyUsedTag) => {
+            const tagObject = sortedTags.find((tag) => tag.name === recentlyUsedTag);
+            return !!tagObject?.enabled && !selectedOptionNames.includes(recentlyUsedTag) && tagObject?.pendingAction !== CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE;
+        })
+        .map((tag) => ({name: tag, enabled: true}));
+
+    if (selectedOptions.length) {
+        tagSections.push({
+            // "Selected" section
+            title: '',
+            shouldShow: true,
+            data: getTagsOptions(selectedTagsWithDisabledState, selectedOptions),
+        });
+    }
+
+    if (filteredRecentlyUsedTags.length > 0) {
+        const cutRecentlyUsedTags = filteredRecentlyUsedTags.slice(0, maxRecentReportsToShow);
+
+        tagSections.push({
+            // "Recent" section
+            title: Localize.translateLocal('common.recent'),
+            shouldShow: true,
+            data: getTagsOptions(cutRecentlyUsedTags, selectedOptions),
+        });
+    }
+
+    tagSections.push({
+        // "All" section when items amount more than the threshold
+        title: Localize.translateLocal('common.all'),
+        shouldShow: true,
+        data: getTagsOptions(enabledTagsWithoutSelectedOptions, selectedOptions),
+    });
+
+    return tagSections;
+}
+
+/**
+ * Verifies that there is at least one enabled tag
+ */
+function hasEnabledTags(policyTagList: Array<PolicyTagLists[keyof PolicyTagLists]>) {
+    const policyTagValueList = policyTagList
+        .filter((tag) => tag && tag.tags)
+        .map(({tags}) => Object.values(tags))
+        .flat();
+
+    return OptionsListUtils.hasEnabledOptions(policyTagValueList);
+}
+
+/**
+ * Sorts tags alphabetically by name.
+ */
+function sortTags(tags: Record<string, PolicyTag | SelectedTagOption> | Array<PolicyTag | SelectedTagOption>) {
+    // Use lodash's sortBy to ensure consistency with oldDot.
```

Comment:
I see this is missing the `const sortedTags = Array.isArray(tags) ? tags : Object.values(tags);` that was in the previous implementation - do we not need that check anymore?